File: 03-exercise/01-solution.hs
-- Übung 03 - Programmierung TU Dresden
-- Tutor: Claas de Boer
-- Github: https://github.com/cdboer/programmierung-ss19


-- Aufgabe 01A
data Tree = Node Int Tree Tree | NIL

insert :: Tree -> [Int] -> Tree
insert t [] = t
insert NIL (x:xs) = insert (Node x NIL NIL) xs
insert (Node a t1 t2) (x:xs)
    | x < a = insert (Node a (insert t1 [x]) t2) xs
    | otherwise = insert (Node a t1 (insert t2 [x])) xs

-- Aufgabe 01B
treeEqual :: Tree -> Tree -> Bool
treeEqual NIL NIL = True
treeEqual NIL (Node x t1 t2) = False
treeEqual (Node x t1 t2) NIL = False
treeEqual (Node x t1 t2) (Node y ta tb) 
    = (x == y) && (treeEqual t1 ta) && (treeEqual t2 tb)

File: 09-exercise/01-solution.pl
% Übung 09 - Programmierung TU Dresden
% Tutor: Claas de Boer
% Github: https://github.com/cdboer/programmierung-ss19

nat(0).
nat(s(X)) :- nat(X).
sum(0, Y, Y) :- nat(Y).
sum(s(X), Y, s(S)) :- sum(X, Y, S).

% Aufgabe 01A
lt(0, s(M)) :- nat(M).
lt(s(N), s(M)) :- lt(N, M).

div(0, M, 0) :- lt(0, M).
div(N, M, 0) :- lt(N, M).
div(N, M, s(Q)) :- lt(0, M), sum(M, V, N), div(V, M, Q).
